man/man3/h2timeLib.3
.\" Copyright (c) 1999-2005 CNRS/LAAS
.\"
.\" Permission to use, copy, modify, and distribute this software for any
.\" purpose with or without fee is hereby granted, provided that the above
.\" copyright notice and this permission notice appear in all copies.
.\"
.\" THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
.\" WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
.\" MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
.\" ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
.\" WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
.\" ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
.\" OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
.\"
.Dd June 23, 2005
.Dt h2timeLib 3
.Os pocoLibs
.Sh NAME 
.Nm h2timeLib
.Nd pocoLibs time and date manipulation library
.Sh SYNOPSIS
.Fd "#include <portLib.h>"
.Fd "#include <h2timeLib.h>"
.Ft void
.Fn h2timeInit void
.Ft STATUS
.Fn h2timeGet "H2TIME *pTimeStr"
.Ft STATUS
.Fn h2GetTimeSpec "H2TIMESPEC *pTimeSpec"
.Ft STATUS
.Fn h2timeInterval "H2TIME *pOldTime" "unsigned long *pNmsec"
.Ft void 
.Fn h2timeShow void
.Sh DESCRIPTION
.Fn h2timeInit
initializes the time handling component of 
.Nm pocolibs. This function is normally called by the global
initialization function
.Xr h2initGlob 3 .
.Pp
.Fn h2timeGet
fills the H2TIME_STR structure pointed to by 
.Fa pTimeStr
with the current time. 
.Pp
.Fn h2GetTimeSpec
fills the H2TIMESPEC_STR structure pointed to by
.Fa pTimeSpec
with the current time. 
.Pp
.Fn h2timeInterval
computes the elapsed number of mili-seconds between the time structure
pointed to by
.Fa pOldTime
and the current time. The value is returned in the long pointed to by
.Fa pNmsec .
.Pp
.Fn h2timeShow
displays the current time and date on the standard output.
.Sh SEE ALSO
.Xr sysLib 3


CHANGELOG
Version 2.8, xx, xxxx, xxxx

104. Nuke RCS ids in source files.
103. Call mlockall() in h2initGlob() when running under Xenomai.
102. Go back to pmap_unset() to unregister pre-existing posterServer service.
101. Fixes for building Xenomai support.
100. Fixes for cross-compilation in OpenEmbedded
 99. Display the ntick field of poster's timestamps in posterShow().
 98. Enforce stack size in taskSpawn() using pthread_attr_setstacksize().

Version 2.7, 16 april 2010

 97. Fix updating 'ntick' field in H2TIME timespamps of posters.
     (fallout from change #94, which broke pom-genom).

Version 2.6, 13 April 2010

 96. On Linux systems, let taskSuspend() throw a backtrace to stderr.
 95. Use a pipe to wait until h2dev lock is ready before launching posterServ.
 94. Implement poster time stamps using H2TIMESPEC instead of H2TIME. 
 93. Don't start posterServ if it cannot find h2 devices.
 92. Add regression tests for h2time conversion functions
 91. Constify posterName arguments in posterLib.
 90. A few more fixes for running modules on Darwin.
 89. Add regression tests for h2 semaphores.
 88. Change h2 semaphores to mutexes instead of counting semaphores.

Version 2.5, 6 April 2009

 87. Update shipped autotools files to the version provided by Fedora 10.
 86. Implement poster access statistics.
 85. posterServ: use svc_unregister() instead of pmap_unset().

Version 2.4.1, 9 September 2008

 84. Fix a stack corruption bug introduced by change 83.

Version 2.4, 11 August 2008

 83. Add a 'taskFromThread(char *name)' function. (Matthieu Gallien).
 83. Assume that target systems have posix timers when cross-compiling
 82. Let autoheader generate pocolibs-config.h.in
 81. Switch repository to git
 80. Fix a race in change #77.
 79. Eliminate more uses of sprintf().
 78. Fix tickLib regression test.

Version 2.3, 31 July 2007

 77. Fix taskDelete hooks handling.
 76. Relax constraint on h2timer periods.
 75. h2 init no longer starts posterServ by default. Use h2 init -p.
 74. Add a flag to h2devInit() to control the launch of posterServ.
     Bump portLib major.
 73. Fix build of posterLib tests programs.
 72. Remove VxWorks/RTAI/LXRT parts.

Version 2.2, 19 December 2006

 71. Let configure find the gnu tar program for make dist.
 70. Remove the xes subsystem, which is no longer used.
 69. Preliminary support for building Xenomai 2.x enabled libs, using the
     POSIX skin.
 68. Remove RTAI/LXRT support in configure scripts
 67. Remove use of X11 events in xes_server.
 66. Remove double clear of h2timer counters (Sylvain Joyeux).
 65. Fix build problems on Darwin.
 64. Provide a correct day of week in H2TIME structure.
 63. Fix h2timeShow() that was broken in multiple ways.

Version 2.1, 11 November 2005

 62. Some basic support for kbuild for RTAI 3.1 kernel modules
     (not working).
 61. Rework error messages handling. No more central h2errorList.
 60. More ANSI C prototypes in header files.
 59. Add a autogen.sh script to generate the configure script and garbage
     collect unneeded files.
 58. Merge missing parts of lxrt support in existing files.
 57. Add missing files for lxrt support.

Version 2.1beta1, 26 August 2005

 56. Use vpath <pattern> instead of VPATH in Makefiles to minimize trouble
     when builddir != srcdir and srcdir has been configured already.
 55. Add a man page for h2time functions.
 54. Add a h2GetTimeSpec() function that returs a timespec-like structure.
     This should be used instead of calling clock_gettime() directly.
 53. Check for clock_gettime in configure scripts
 52. Add regression tests.
 51. Packaging fixes.
 50. Add a pkg-config data file and install it
 49. Write manual pages for portLib
 48. return S_symLib_NOT_IMPLEMENTED for symFindByValue()
 47. sprintf() -> snprintf() conversion
 46. less printf()s in portLib
 45. more comments translation
 44. Makefiles cleanups (Sylvain Joyeux, Matthieu Herrb).
 43. Add a h2semSet() function to force the value of a semaphore (Sylvain
     Joyeux).
 42. Add a posterName() function and a FIO_FRESH ioctl to posterLib (Sylvain
     Joyeux).
 41. Valgrind support (Sylvain Joyeux).
 40. Remove some unused files.
 39. Make headers C++ compatible (Sylvain Joyeux).
 38. Add --{disable,enable}-posix-timer option to configure, to be able
     to force the timer type that is used by tickLib, as the auto-detection
     of useable POSIX timer is not always reliable enough.
 37. Added --with-linux-kernel-includes option to specify the path to
     linux kernel headers when compiling for RTAI.
 36. Revert the change that made the posterId in posterServ an uint64_t.
 35. Autotoolize the build of the poster test programs.
 34. Start over with libtool style shared libraries version numbers.
     Bump all major revisions to make sure nothing bad happens.
 33. Add a configure test for usabilility of POSIX timers for a
     100Hz sysClk.
 32. Add a 'tick' regress test to check the accuracy of sysClk.
     (Linux 2.4 POSIX timers are known to be unable to set a clock
     at 100Hz).
 31. Some cleanups in header files: avoid nested includes.
 30. Fix posterLib LP64 issues.
 29. Implemented a real-time logging library for RTAI. Added logInit and
     logEnd. Bumped portLib minor.
 28. Updated RTAI/user-space to handle multi-threaded linux processes.
     It now uses one real-time task per thread.
 27. Added devfs support for RTAI.
 26. install-sh does not overwrite files if they were not modified.
 25. Preliminary RTAI/user-space support. posterLib requires a more
     complicated implementation: as of now, posters can be shared between
     RTAI-tasks and linux processes but this breaks the real-time
     properties of RTAI-tasks.
 24. Added taskDelete() regression test.
 23. Added --with-debug option to configure script. It allows the
     selection of independant components to debug.
 22. Added h2devClean() in comLib. This was previously defined in h2.c
     but it was problematic for user-space RTAI version. Bumped comLib

src/comLib/h2timeLib.c
/*
 * Copyright (c) 1990, 2003-2004 CNRS/LAAS
 *
 * Permission to use, copy, modify, and distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 */

#include "pocolibs-config.h"


#include <stdio.h>
#include <string.h>
#include <time.h>
#include <sys/time.h>
#include <sys/types.h>

#include "portLib.h"
#include "h2timeLib.h"

static H2TIMESPEC h2TimeSpec0;
static void h2timespec_subtract(H2TIMESPEC *,
    const H2TIMESPEC *, const H2TIMESPEC *);

/*----------------------------------------------------------------------*/

/**
 **  h2timeGet  -  Lire la valeur actuelle de l'horloge
 **
 **  Description :
 **  Lit la valeur actuelle de l'horloge et la met dans la structure
 **  fournit par l'utlisateur.
 **
 **  Retourne : OK ou ERROR
 **/

STATUS
h2timeGet(H2TIME *pTimeStr)
{
    struct timeval tv;

    gettimeofday(&tv, NULL);

    h2timeFromTimeval(pTimeStr, &tv);
    /* free(tmp); */

    return(OK);
} /* h2timeGet */

STATUS
h2GetTimeSpec(H2TIMESPEC *pTs)
{
#ifdef HAVE_CLOCK_GETTIME
	struct timespec ts;
	if (clock_gettime(CLOCK_REALTIME, &ts) < 0)
		return ERROR;
	if (pTs != NULL) {
		pTs->tv_sec = ts.tv_sec;
		pTs->tv_nsec = ts.tv_nsec;
	}
#else
	struct timeval tv;

	gettimeofday(&tv, NULL);
	if (pTs != NULL) {
		pTs->tv_sec = tv.tv_sec;
		pTs->tv_nsec = tv.tv_usec * 1000;
	}
#endif
	return OK;
}

static int days_in_year[] =
    { 0,                // Janvier
      31,               // Fevrier
      31+28,            // Mars
      2*31+28,          // Avril
      2*31+30+28,       // Mai
      3*31+30+28,       // Juin
      3*31+2*30+28,     // Juillet
      4*31+2*30+28,     // Aout
      5*31+2*30+28,     // Septembre
      5*31+3*30+28,     // Octobre
      6*31+3*30+28,     // Novembre
      6*31+4*30+28,     // Decembre
      7*31+4*30+28
    };
#define is_bisextile_exception(y) ( ((y) % 100 == 0) && ((y) % 400 != 0) )
#define is_bisextile(y)     ( ( !((y) % 4) && !is_bisextile_exception(y) ) ? 1 : 0 )
#define days_since_zero(y)  ( (y) * days_in_year[12] + ((y) / 4) + ((y) / 400) - ((y) / 100) )
#define days_since_epoch(y) ( days_since_zero(y) - days_since_zero(1970) )

void
h2timeFromTimeval(H2TIME* pTimeStr, const struct timeval* tv)
{
    long sec, day, dow, month, year;
    static int day_per_month[] =
	{ 31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31 };

    pTimeStr->ntick = tv -> tv_sec * NTICKS_PER_SEC + tv -> tv_usec / TICK_US;
    pTimeStr->msec = tv -> tv_usec / 1000;

    sec = tv -> tv_sec;
    day  = sec / (3600 * 24); // day since 1970
    dow = (day + 4) % 7;		/* the 1/1/70 was a Thurday */
    year = (day / 365 + 1970);        // We'll have one year difference each 4*365 years (roughly)
    day  -= days_since_epoch(year);  // day in the year

    if (day < 0) {
	    day = 365 + day;
	    year -= 1;
    }
    day_per_month[1] = is_bisextile(year) ? 29 : 28;
    for (month = 0; month < 12; ++month)
    {
	if (day < day_per_month[month])
	    break;
	day -= day_per_month[month];
    }
    // now, day is the day in the month and month the month index

    pTimeStr->sec    = sec % 60;
    pTimeStr->minute = (sec / 60) % 60;
    pTimeStr->hour   = (sec / 3600) % 24;
    pTimeStr->day   = dow;
    pTimeStr->date  = day + 1;
    pTimeStr->month = month + 1;
    pTimeStr->year  = year - 1900;
//#else
//    tmp = localtime((time_t *)&tv -> tv_sec);
//
//    pTimeStr->sec    = tmp->tm_sec;
//    pTimeStr->minute = tmp->tm_min;
//    pTimeStr->hour   = tmp->tm_hour;
//    pTimeStr->day    = tmp->tm_wday == 0 ? 7 : tmp->tm_wday;
//    pTimeStr->date   = tmp->tm_mday;
//    pTimeStr->month  = tmp->tm_mon + 1;
//    pTimeStr->year   = tmp->tm_year;
//#endif
}

void
timevalFromH2time(struct timeval* tv, const H2TIME* pTimeStr)
{
    int year = pTimeStr->year + 1900;
    int days = pTimeStr->date - 1
	     + days_in_year[pTimeStr->month - 1]
	     + ((pTimeStr->month > 2) ? is_bisextile(year) : 0)
	     + days_since_epoch(year);

    tv -> tv_usec = pTimeStr->msec*1000;

    tv -> tv_sec
	= pTimeStr->sec
	    + pTimeStr->minute*60
	    + pTimeStr->hour  *3600
	    + days            *24*3600;
}

/*----------------------------------------------------------------------*/
void
h2timeFromTimespec(H2TIME* pTimeStr, const H2TIMESPEC *ts)
{
	struct tm tm;
	H2TIMESPEC diff;
	unsigned long rate = sysClkRateGet();

	h2timespec_subtract(&diff, ts, &h2TimeSpec0);
	gmtime_r(&ts->tv_sec, &tm);

	pTimeStr->ntick  = diff.tv_sec*rate
	    + (unsigned long long)diff.tv_nsec*rate/1000000000ULL;
	pTimeStr->msec   = diff.tv_nsec / 1000000;
	pTimeStr->sec    = tm.tm_sec;
	pTimeStr->minute = tm.tm_min;
	pTimeStr->hour   = tm.tm_hour;
	pTimeStr->day    = tm.tm_wday == 0 ? 7 : tm.tm_wday;
	pTimeStr->date   = tm.tm_mday;
	pTimeStr->month  = tm.tm_mon + 1;
	pTimeStr->year   = tm.tm_year;
}

/*----------------------------------------------------------------------*/
void
timespecFromH2time(H2TIMESPEC *ts, const H2TIME *pTimeStr)
{
	struct tm tm;

	tm.tm_sec = pTimeStr->sec;
	tm.tm_min = pTimeStr->minute;
	tm.tm_hour = pTimeStr->hour;
	tm.tm_wday = pTimeStr->day;
	tm.tm_mday = pTimeStr->date;
	tm.tm_mon = pTimeStr->month - 1;
	tm.tm_year = pTimeStr->year;

	ts->tv_sec = mktime(&tm);
	ts->tv_nsec = pTimeStr->msec*1000000;
}
/*----------------------------------------------------------------------*/

STATUS
h2timeSet(H2TIME *pTimeStr)
{

    return(ERROR);
} /* h2timeSet */

/*----------------------------------------------------------------------*/

STATUS
h2timeAdj(H2TIME *pTimeStr)
{

    return(ERROR);
} /* h2timeAdj */

/*----------------------------------------------------------------------*/

/**
 **  h2timeShow  - Envoyer la date vers la sortie standard
 **
 **  Description:
 **  Envoie la date vers la sortie standard.
 **
 **  Retourne : Rien
 **/
void
h2timeShow(void)
{
    H2TIME strTime;
    static char *dayStr [] = {
	    "sunday", "monday", "tuesday", "wednesday", "thursday",
	    "friday", "saturday"};
    char buf[50];

    /* Demander la lecture de la date */
    if (h2timeGet (&strTime) != OK) {
	logMsg ("Problem reading date/time!\n");
	return;
    }

    /* Envoyer la date vers la console */
    snprintf(buf, sizeof(buf),
	"\nDate: %02u-%02u-%02u, %s, %02uh:%02umin:%02us\n\n",
	strTime.month, strTime.date, strTime.year + 1900,
	dayStr[strTime.day], strTime.hour, strTime.minute,
	strTime.sec);
    logMsg(buf);
} /* h2timeShow */

/*----------------------------------------------------------------------*/

/**
 **  h2timeInterval  -  Intervalle de temps entre deux lectures horloge
 **
 **  Description:
 **  Donne l'ecart en milli-secondes entre une ancienne lecture de l'horloge
 **  et sa valeur actuelle.
 **
 **  Retourne : OK ou ERROR
 **/

STATUS
h2timeInterval (H2TIME *pOldTime, unsigned long *pNmsec)
{
    H2TIME strTime;		/* Ou` mettre la valeur actuelle */
    unsigned long ntick1, ntick2;	/* Nombre de ticks */

    /* Lire l'horloge */
    if (h2timeGet (&strTime) != OK)
	return (ERROR);

    /* Charger les compteurs de ticks */
    ntick2 = strTime.ntick;
    ntick1 = pOldTime->ntick;

    /* Retourner l'intervalle de temps */
    *pNmsec = 1000 * ((ntick2 >= ntick1) ? (ntick2 - ntick1)
		   : (((unsigned long) ~0 - ntick1) + 1 + ntick2))
			/NTICKS_PER_SEC;
    return (OK);
} /* h2timeInterval */

/*----------------------------------------------------------------------*/

/**
 ** Returns the difference in mili-seconds between current time and
 ** an older H2TIMESPEC value
 **/

STATUS
h2timespecInterval(const H2TIMESPEC *pOldTime, unsigned long *pNmsec)
{
	H2TIMESPEC ts, res;

	if (h2GetTimeSpec(&ts) != OK)
		return (ERROR);
	h2timespec_subtract(&res, &ts, pOldTime);
	*pNmsec = 1000*res.tv_sec + res.tv_nsec / 1000000;
	return OK;
}

/*----------------------------------------------------------------------*/

static void
h2timespec_subtract(H2TIMESPEC *result,
    const H2TIMESPEC *x, const H2TIMESPEC *y)
{
	H2TIMESPEC yy;

	memcpy(&yy, y, sizeof(H2TIMESPEC));

	/* Carry */
	if (x->tv_nsec < y->tv_nsec) {
		long sec = (y->tv_nsec - x->tv_nsec) / 1000000000 + 1;
		yy.tv_nsec -= 1000000000 * sec;
		yy.tv_sec += sec;
	}
	if (x->tv_nsec - y->tv_nsec > 1000000000) {
		int sec = (x->tv_nsec - y->tv_nsec) / 1000000000;
		yy.tv_nsec += 1000000000 * sec;
		yy.tv_sec -= sec;
	}

	result->tv_sec = x->tv_sec - yy.tv_sec;
	result->tv_nsec = x->tv_nsec - yy.tv_nsec;
}

/*----------------------------------------------------------------------*/

void
h2timeInit(void)
{
	h2GetTimeSpec(&h2TimeSpec0);
	h2TimeSpec0.tv_sec = h2TimeSpec0.tv_sec - h2TimeSpec0.tv_sec % 86400;
	h2TimeSpec0.tv_nsec = 0;
} /* h2timeInit */

/*----------------------------------------------------------------------*/
